client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -150,20 +150,19 @@ private PositionInternal getPosition() {
     @GuardedBy("readers")
     private String updateGroupStateIfNeeded() throws ReinitializationRequiredException {
```

Comment:
I'm confused about the final logic of this method. The final code after the changes looks like:

```
            if (atCheckpoint != null) {
                groupState.checkpoint(atCheckpoint, getPosition());
                releaseSegmentsIfNeeded();
                atCheckpoint = null;
            }
            String checkpoint = groupState.getCheckpoint();
            if (checkpoint != null) {
                log.info("{} at checkpoint {}", this, checkpoint);
                atCheckpoint = checkpoint;
                return checkpoint;
            }
```

In this flow, we check whether `atCheckpoint` is null, followed by getting a checkpoint name and setting `atCheckpoint` in the case there is an ongoing checkpoint. Shouldn't we execute the `atCheckpoint != null` in the case that we are setting `atCheckpoint` rather than wait until the next call to this method?

Reply:
No. That change in flow is the purpose of the PR. (See the issue)

Reply:
I believe I understand the issue, so let me go with that assumption. The problem I'm seeing is that there is nothing in these changes that prevents a call to `updateGroupStateIfNeeded()` before `readNextEvent` is invoked next, so there is no effective difference between executing the first if block above after the second sets `atCheckpoint` at least I'm not sure how we guarantee that `updateGroupStateIfNeeded()` won't be called again until `readNextEvent` is invoked next. 

What am I missing?

Reply:
See if the updated Javadoc clarifies this.